### PR TITLE
Add i18n to topology package

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
Fixes: https://issues.redhat.com/browse/ODC-5130

### CODE REVIEW

File: frontend/packages/topology/src/utils/moveNodeToGroup.tsx [hunk 1]
```diff
@@ -1,21 +1,31 @@
...
+        {targetGroup
+          ? t('topology~Are you sure you want to move ')
+          : t('topology~Are you sure you want to remove ')}
+        <strong>{node.getLabel()}</strong>
+        {targetGroup
+          ? t('topology~ from {{sourceGroup}} to  {{targetGroup}}?', {
+              sourceGroup: sourceGroup.getLabel(),
+              targetGroup: targetGroup.getLabel(),
+            })
+          : t('topology~ from {{sourceGroup}}?', { sourceGroup: sourceGroup.getLabel() })}
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/utils/moveNodeToGroup.tsx:28]: 🤔  There is a pattern that can be used here...

Quick search through the codebase got me: 
```
"teams-help": "Optionally list organizations. If specified, only GitHub users that are members of at least one of the listed organizations will be allowed to log in. Cannot be used in combination with <1>teams</1>.",
```
And:
```
            <Trans i18nKey="teams-help" ns="github-idp-form">
              Optionally list organizations. If specified, only GitHub users that are members of at
              least one of the listed organizations will be allowed to log in. Cannot be used in
              combination with <strong>teams</strong>.
            </Trans>
```
Ref: [github-idp-form.tsx](https://github.com/openshift/console/blob/master/frontend/public/components/cluster-settings/github-idp-form.tsx#L228-L232)

cc @christianvogt might know more -- I'm reverse engineering here.

File: frontend/packages/topology/src/actions/groupActions.ts [hunk 2]
```diff
@@ -29,15 +30,15 @@ const cleanUpWorkloadNode = async (workload: OdcNodeModel): Promise<K8sResourceK
   return cleanUpWorkload(workload.resource, workload.data?.isKnativeResource ?? false);
 };
 
-const deleteGroup = (application: TopologyApplicationObject) => {
+const deleteGroup = (application: TopologyApplicationObject, t: TFunction) => {
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/actions/groupActions.ts:33]: Interesting, you put `t` at the start of other methods 🤔 

I have no opinion one way or another, just thought it was noteworthy.

File: frontend/packages/topology/src/components/modals/MoveConnectionModal.tsx [hunk 3]
```diff
@@ -70,12 +74,12 @@ const MoveConnectionForm: React.FC<FormikProps<FormikValues> & MoveConnectionMod
...
     <form onSubmit={handleSubmit} className="modal-content modal-content--no-inner-scroll">
-      <ModalTitle>Move Connector</ModalTitle>
+      <ModalTitle>{t('topology~Move Connector')}</ModalTitle>
       <ModalBody>
         <Title headingLevel="h2" size="md" className="co-m-form-row">
-          Connect
+          {t('topology~Connect')}
           <strong>{` ${edge.getSource().getLabel()} `}</strong>
-          to
+          {t('topology~to')}
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/components/modals/MoveConnectionModal.tsx:82]: This should be one translation otherwise it won't come out properly in other languages.

File: frontend/packages/topology/src/operators/TopologyServiceBindingRequestPanel.tsx [hunk 4]
```diff
@@ -16,14 +19,31 @@ type TopologyServiceBindingRequestPanelProps = {
...
+  selectedDetailsTab: UI.getIn(['overview', 'selectedDetailsTab']),
+});
+
+const dispatchToProps = (dispatch): PropsFromDispatch => ({
+  onClickTab: (name) => dispatch(UIActions.selectOverviewDetailsTab(name)),
+});
+
+const ConnectedTopologyServiceBindingRequestPanel: React.FC<PropsFromState &
+  PropsFromDispatch &
+  TopologyServiceBindingRequestPanelProps> = ({ edge, onClickTab, selectedDetailsTab }) => {
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/operators/TopologyServiceBindingRequestPanel.tsx:40]: Not sure I understand what is happening here... why this change? Seems like new functionality for an I18n PR.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

**Comment by andrewballantyne** [comment]
Add page looks translated...
Screen Shot 2020-11-26 at 2 48 34 PM
But we have actions that don't look translated:
Screen Shot 2020-11-26 at 2 48 21 PM
Screen Shot 2020-11-26 at 2 49 13 PM
Screen Shot 2020-11-26 at 2 52 31 PM

---

Sidebar has a lot of non-translated chunks... not sure if that's a topology package issue with the new code location 🤔 
Screen Shot 2020-11-26 at 2 49 56 PM

---

All the actions but the two layouts look uncaught by I18n...
Screen Shot 2020-11-26 at 2 50 11 PM

---

Partially missing translation:
Screen Shot 2020-11-26 at 2 51 23 PM

---

Not sure who covers the build decorator 🤔  This is in the Topology package... so I assume this PR...
Screen Shot 2020-11-26 at 2 52 08 PM

File: frontend/packages/topology/src/utils/removeConnection.tsx [hunk 5]
```diff
@@ -19,10 +21,10 @@ export const removeConnection = (edge: Edge): Promise<any> => {
       </>
     ),
     message,
-    btnText: 'Delete',
+    btnText: t('topology~Delete'),
     submitDanger: true,
     executeFn: () => {
-      return removeTopologyResourceConnection(edge).catch((err) => {
+      return removeTopologyResourceConnection(t, edge).catch((err) => {
```
Comment (christianvogt) [inline frontend/packages/topology/src/utils/removeConnection.tsx:27]: I'd prefer to have `confirmModal` accept a callback which is either executed in the render context so that the function can use `useTranslation` or it gets passed in the translation function. This would make it easier to use these modal launchers without the need for the consumer to provide it's own `t` function.

File: frontend/packages/topology/src/utils/topology-utils.ts [hunk 6]
```diff
@@ -97,6 +98,7 @@ export const getResourceKind = (node: Node): K8sResourceKindReference => {
 };
 
 export const updateTopologyResourceApplication = (
+  t: TFunction,
```
Comment (christianvogt) [inline frontend/packages/topology/src/utils/topology-utils.ts:101]: Again it seems we're passing around `t` to localize the return value of async operations that won't be re-run if the locale changes.
Seems unnecessary to pass `t` around everywhere for this and instead use the global `i18next`

File: frontend/packages/topology/src/components/page/TopologyPageToolbar.tsx [hunk 7]
```diff
@@ -25,8 +27,8 @@ const TopologyPageToolbar: React.FC<TopologyPageToolbarProps> = observer(
       <>
         {showGraphView ? (
           <Popover
-            aria-label="Shortcuts"
-            bodyContent={TopologyShortcuts}
+            aria-label={t('topology~Shortcuts')}
+            bodyContent={getTopologyShortcuts(t)}
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/page/TopologyPageToolbar.tsx:31]: The API accepts a `ReactNode`. `TopologyShortcuts` can make use of `useTranslation`
```suggestion
            bodyContent={<TopologyShortcuts />}
```

File: frontend/packages/topology/src/components/modals/MoveConnectionModal.tsx [hunk 8]
```diff
@@ -42,6 +45,7 @@ const MoveConnectionForm: React.FC<FormikProps<FormikValues> & MoveConnectionMod
   isSubmitting,
   cancel,
   values,
+  t,
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/modals/MoveConnectionModal.tsx:48]: Use `useTranslation` instead.

File: frontend/packages/topology/src/components/modals/EditApplicationModal.tsx [hunk 9]
```diff
@@ -31,6 +34,7 @@ type EditApplicationModalProps = EditApplicationFormProps & {
 };
 
 const EditApplicationForm: React.FC<FormikProps<FormikValues> & EditApplicationFormProps> = ({
+  t,
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/modals/EditApplicationModal.tsx:37]: Use `useTranslation` instead.

File: frontend/packages/topology/src/components/graph-view/components/nodeContextMenu.tsx [hunk 10]
```diff
@@ -75,7 +76,7 @@ export const groupContextMenu = (element: Node, connectorSource?: Node) => {
 
   const graphData = element.getGraph().getData();
   return createMenuItems(
-    kebabOptionsToMenu(groupActions(graphData, applicationData, connectorSource)),
+    kebabOptionsToMenu(groupActions(t, graphData, applicationData, connectorSource)),
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/graph-view/components/nodeContextMenu.tsx:79]: Actions can use `labelKey`

**Review by christianvogt — COMMENTED** [review]
Overall I question the use of passing around `t` over simply using the global translation function when the translation of errors isn't reactive anyways.

**Comment by jeff-phillips-18** [comment]
@andrewballantyne Side bar is part of public/components/overview (not topology). I believe I have addressed other comments.

File: frontend/packages/topology/src/filters/TopologyFilterBar.tsx [hunk 11]
```diff
@@ -53,7 +54,11 @@ const TopologyFilterBar: React.FC<TopologyFilterBarProps> = ({
   viewType,
   namespace,
 }) => {
+<<<<<<< HEAD
   const { filters, setTopologyFilters: onFiltersChange } = React.useContext(FilterContext);
+=======
+  const { t } = useTranslation();
+>>>>>>> Add i18n to topology package
```
Comment (vikram-raj) [inline frontend/packages/topology/src/filters/TopologyFilterBar.tsx:61]: @jeff-phillips-18 Missed this Merge conflict marker encountered.

File: frontend/packages/topology/src/utils/moveNodeToGroup.tsx [hunk 12]
```diff
@@ -10,15 +11,29 @@ export const moveNodeToGroup = (node: Node, targetGroup: Node): Promise<void> =>
...
-    const message = (
-      <>
-        Are you sure you want to {targetGroup ? 'move' : 'remove'}{' '}
-        <strong>{node.getLabel()}</strong> from {sourceGroup.getLabel()}
-        {targetGroup ? ` to ${targetGroup.getLabel()}` : ''}?
-      </>
+    const title = targetGroup ? (
+      <Trans ns="topology">Move Component Node</Trans>
+    ) : (
+      <Trans ns="topology">Remove Component Node from Application</Trans>
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/utils/moveNodeToGroup.tsx:17]: Don't think you need to use `Trans` on static strings 🤔

File: frontend/packages/topology/src/utils/moveNodeToGroup.tsx [hunk 13]
```diff
@@ -10,15 +11,29 @@ export const moveNodeToGroup = (node: Node, targetGroup: Node): Promise<void> =>
...
+      </Trans>
+    ) : (
+      <Trans ns="topology">
+        Are you sure you want to remove <strong>{{ nodeLabel }}</strong> from {{ sourceLabel }}?
+      </Trans>
+    );
+    const btnText = targetGroup ? (
+      <Trans ns="topology">Move</Trans>
+    ) : (
+      <Trans ns="topology">Remove</Trans>
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/utils/moveNodeToGroup.tsx:35]: Same here

File: frontend/packages/topology/src/utils/removeConnection.tsx [hunk 14]
```diff
@@ -7,19 +8,22 @@ import { removeTopologyResourceConnection } from './topology-utils';
 export const removeConnection = (edge: Edge): Promise<any> => {
   const message = (
     <p>
-      Deleting the visual connector removes the `connects-to` annotation from the resources. Are you
-      sure you want to delete the visual connector?
+      <Trans ns="topology">
+        Deleting the visual connector removes the `connects-to` annotation from the resources. Are
+        you sure you want to delete the visual connector?
+      </Trans>
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/utils/removeConnection.tsx:14]: Same here

File: frontend/packages/topology/src/utils/removeConnection.tsx [hunk 15]
```diff
@@ -7,19 +8,22 @@ import { removeTopologyResourceConnection } from './topology-utils';
...
     </p>
   );
 
   return confirmModal({
     title: (
       <>
-        <YellowExclamationTriangleIcon className="co-icon-space-r" /> Delete Connector?
+        <YellowExclamationTriangleIcon className="co-icon-space-r" />{' '}
+        <Trans ns="topology">Delete Connector?</Trans>
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/utils/removeConnection.tsx:22]: Same here

File: frontend/packages/topology/src/utils/removeConnection.tsx [hunk 16]
```diff
@@ -7,19 +8,22 @@ import { removeTopologyResourceConnection } from './topology-utils';
...
       </>
     ),
     message,
-    btnText: 'Delete',
+    btnText: <Trans ns="topology">Delete</Trans>,
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/utils/removeConnection.tsx:26]: Same here

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Not sure I misunderstand `Trans` vs `t` ... but I think you can use `t` for all strings, and `Trans` when React components or DOM elements get in the way

File: frontend/packages/topology/src/components/side-bar/TopologyEdgeResourcesPanel.tsx [hunk 17]
```diff
@@ -17,6 +18,7 @@ type TopologyEdgeResourcesPanelProps = {
 };
 
 const TopologyEdgeResourcesPanel: React.FC<TopologyEdgeResourcesPanelProps> = ({ edge }) => {
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/components/side-bar/TopologyEdgeResourcesPanel.tsx:20]: Missed `Connections` string passed to `SidebarSectionHeading` on line 36.

Comment (andrewballantyne) [reply frontend/packages/topology/src/components/side-bar/TopologyEdgeResourcesPanel.tsx:20]: Oooh, missed the title missing. 

With the pseudo lang:
Screen Shot 2020-12-01 at 11 36 18 AM

With the "default" lang:
Screen Shot 2020-12-01 at 11 36 23 AM

**Review by vikram-raj — COMMENTED** [review]
missing header of the connector sidepanel
missing-connector-header

missing key warning
missing-key-warning

**Comment by jeff-phillips-18** [comment]
@andrewballantyne  @vikram-raj Updated per comments.

File: frontend/packages/topology/src/components/list-view/cells/StatusCell.tsx [hunk 18]
```diff
@@ -50,11 +53,11 @@ const StatusCellResourceStatus: React.FC<StatusCellResourceStatus> = ({ obj, pod
     if (!filteredPods.length) {
       return null;
     }
-    return <Link to={href}>{filteredPods.length} pods</Link>;
+    return <Link to={href}>{t('topology~{{length}} pods', { length: filteredPods.length })}</Link>;
```
Comment (vikram-raj) [inline frontend/packages/topology/src/components/list-view/cells/StatusCell.tsx:56]:
```suggestion
    return <Link to={href}>{t('topology~{{count, number}} pod', { count: filteredPods.length })}</Link>;
```

**Review by andrewballantyne — APPROVED** [review]
/lgtm
